File: 15/b.ts
// const input = await Bun.file("test.txt").text();
const input = await Bun.file("input.txt").text();

const hash = (line: string) => {
  let acc = 0;
  for (const char of line) {
    acc += char.charCodeAt(0);
    acc *= 17;
    acc %= 256;
  }
  return acc;
};

let map: [string, number][][] = [];

input.split(",").forEach((line) => {
  if (line.includes("=")) {
    let [label, focalLength] = line.split("=");
    let n = hash(label);
    if (!map[n]) map[n] = [];
    let idx = map[n].findIndex((pair) => pair[0] === label);
    if (idx !== -1) {
      map[n][idx][1] = parseInt(focalLength);
    } else {
      map[n].push([label, parseInt(focalLength)]);
    }
  } else {
    let label = line.slice(0, -1);
    let n = hash(label);
    if (!map[n]) map[n] = [];
    map[n] = map[n].filter((pair) => pair[0] !== label);
  }
});

let res = 0;
map.forEach((arr, i) => {
  arr.forEach((pair, j) => {
    res += (i + 1) * (j + 1) * pair[1];
  });
});

console.log(res);

File: 15/b.go
package main

import (
	"fmt"
	"os"
	"slices"
	"strings"
)

type Lens struct {
	label       string
	focalLength int
}

func main() {
	file, _ := os.ReadFile("input.txt")

	boxes := make(map[int][]Lens)

	for _, str := range strings.Split(string(file), ",") {
		hash := 0
		label := ""
		isFocalLength := false
		focalLength := 0
		for _, char := range str {
			if char == '-' {
				if boxes[hash] == nil {
					boxes[hash] = make([]Lens, 0)
				}
				for i, lens := range boxes[hash] {
					if lens.label == label {
						boxes[hash] = slices.Delete(boxes[hash], i, i+1)
						break
					}
				}
				break
			}
			if char == '=' {
				isFocalLength = true
				continue
			}
			if isFocalLength {
				focalLength = focalLength*10 + int(char-'0')
				continue
			}
			label += string(char)
			hash += int(char)
			hash *= 17
			hash %= 256
		}
		if !isFocalLength {
			continue
		}
		lens := Lens{label, focalLength}
		foundDuplicate := false
		for i, l := range boxes[hash] {
			if l.label == label {
				boxes[hash][i] = lens
				foundDuplicate = true
				break
			}
		}
		if !foundDuplicate {
			boxes[hash] = append(boxes[hash], lens)
		}
		// fmt.Println(boxes)
	}

	res := 0
	for i, lenses := range boxes {
		for j, lens := range lenses {
			res += lens.focalLength * (i + 1) * (j + 1)
		}
	}

	fmt.Println(res)
}
